proposals/ideas/xxx-padding-machines-onions.txt

```diff
@@ -0,0 +1,302 @@
+
+   In the following sections we will analyze these machines.
+
+5.1. Client-side introduction circuit hiding machines [INTRO_CIRC_HIDING]
+
+   These two machines are meant to hide client-side introduction circuits. The
+   origin-side machine sits on the client and sends padding towards the
+   introduction circuit, whereas the relay-side machine sits on the middle-hop
+   (second hop of the circuit) and sends padding towards the client. The
+   padding from the origin-side machine terminates at the middle-hop and does
+   not get forwarded to the actual introduction point.
+
+   Both of these machines only get activated for introduction circuits, and
+   only after an INTRODUCE1 cell has been sent out.
+
+   This means that before the machine gets activated our cell flow looks like this:
+
+    [EXTEND2] -> EXTENDED2 -> [EXTEND2] -> EXTENDED2 -> [EXTEND2] -> EXTENDED2 -> [INTRODUCE1]
+
+   Comparing the above with section [CIRCCONSTRUCTION], we see that the above
+   cell sequence matches the one from general circuits up to the first 7 cells.
+
+   However, in normal introduction circuits this is followed by an
+   INTRODUCE_ACK and then the circuit gets teared down, which does not match
+   the sequence from [CIRCCONSTRUCTION].
+
+   Hence when our machine is used, after sending an [INTRODUCE1] cell, we also
+   send a [PADDING_NEGOTIATE] cell, which gets answered by a PADDING_NEGOTIATED
+   cell and an INTRODUCE_ACKED cell. This makes us match the [CIRCCONSTRUCTION]
+   sequence up to the first 10 cells.
+
+   After that, we continue sending padding from the relay-side machine so as to
+   fake a directory download, or an SSL connection setup. We also want to
+   continue sending padding so that the connection stays up longer to destroy
+   the "Duration of Activity" fingerprint.
+
+   To calculate the padding overhead, we see that the origin-side machine just
+   sends a single [PADDING_NEGOATIATE] cell, wheras the origin-side machine
+   sends a PADDING_NEGOTIATED cell and between 7 to 10 DROP cells. This means
+   that the average overhead of this machine is 11 padding cells.
+
+   In terms of WTF-PAD terminology, these machines have three states (START,
+   BURST, END). They move from the START to BURST state when the first
+   non-padding cell is received on the circuit, and they stay in the BURST
+   state until all the padding gets depleted. The BURST state is controlled by
+   a histogram which specifies the parameters described in the paragraphs
+   above. After all the padding finishes, it moves to END state.
+
+   We also set a special WTF-PAD flag which keeps the circuit open even after
+   the introduction is performed, and until all the padding has been sent
+   out. In particular, the origin-side machine needs in average 3.3 secs to
+   deplete all its padding cells, and hence it will slightly increase the
+   circuit lifetime. See section [IMPROVEMENTS] for a further low-hanging-fruit
+   improvement in this direction.
```

Comment:
Where is the IMPROVEMENTS section?

Also, the manage_circ_lifetime field should keep the circuit open for MaxCircuitDirtiness seconds now. Is that not the case?

Reply:
Oops, the `IMPROVEMENTS` section was removed. It was suggesting to split the machine into multiple phases, so that we try to hide the post-circuit-construction traffic as well.

Good point that the #28780 also messes with `MaxCircuitDiriness` need to update that.